### DISCUSSION TITLE
Numpy/Xtensor like library for C#?

### DISCUSSION THREAD

**Discussion by Kukuschi** [MonoGame/MonoGame#8462] [General]
Hello, 
I am new here and search for the best solution for the game I want to work on next. 
I have experience with some engines (unreal and godot) but am only a mediocre coder (yet). 
What I need for my project would be a fast method to manipulate large amount of voxel data (mostly rgba8) in multidimensional arrays. 

I started the project in Godot, (but I want to use a framework instead for many many different reasons) using the new compositor effects in 4.3, handling the voxel data manipulation on the GPU via compute shader (which comes with it's own problems, especially the very limited GPU RAM).
I wrote a voxel art plugin for Blender3D some time ago and made heavy use of numpy there. I would like to try a similar approach for my game and I stumbled over the Numsharp library and got exited first, but it is more or less incomplete and dead it seems.
Are there any good libraries that could provide me with numpy or xtensor like array manipulation for C#? (My last resort would be to use a c++ framework like SFML, I would much prefer Monogame tho - using c++ seems to add a lot of complexity and work and a much steeper learning curve).

**Comment by stromkos** [comment]
Welcome.

Numpy is written in C++.  Many of the singleton functions are already available in C#. 

You could try C# Linq functions for multi-value processing.

For C#, I recommend learning array pointers in the unsafe context. Not as fast as C++ but close enough.
